src/common/ceph_mutex.h

```diff
@@ -0,0 +1,135 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+// What and why
+// ============
+//
+// For general code making use of mutexes, use these ceph:: types.  The main
+// difference is that the ceph::mutex ctor takes a string argument naming
+// the lock for use by lockdep.
+//
+// For legacy Mutex users that passed recursive=true, use ceph::recursive_mutex.
+//
+// For legacy Mutex users that passed lockdep=false, use std::mutex directly.
+//
+// ceph::unique_lock is meant to replace std::lock_guard<std::mutex>, so as
+// to avoid the fact that ceph::mutex may be something other than std::mutex.
+// It's important to use this type instead of std::lock_guard<ceph::mutex>
+// because std::condition_variable (aka ceph::condition_variable, if debug types
+// are off) *needs* std::unique_lock<std::mutex> because of various
+// optimizations it makes.
+
+#ifdef CEPH_DEBUG_MUTEX
+
+// ============================================================================
+// debug (lockdep-capable, various sanity checks and asserts)
+// ============================================================================
+
+#include "common/mutex_debug.h"
+#include "common/condition_variable_debug.h"
+
+namespace ceph {
+  typedef ceph::mutex_debug mutex;
+  typedef ceph::mutex_recursive_debug recursive_mutex;
+  typedef ceph::condition_variable_debug condition_variable;
+  typedef std::unique_lock<mutex> unique_lock;
```

Comment:
i don't think we'd need this alias. as `std::unique_lock` is a template class. and with C++17, its template argument can be deduced under most circumstances.

Reply:
The problem is it will deduce to std::unique_lock<ceph::mutex> and not std::unique_lock<std::mutex>, which means we won't be able to use std::condition_variable.

Reply:
i don't follow. if `CEPH_DEBUG_MUTEX` is defined. `ceph::mutex` is an alias of `ceph::mutex_debug`. and `ceph::condition_variable` will be `ceph::condition_variable_debug`.

if we want to use `condition_variable` along with a `mutex`, we should be using `ceph::condition_variable`, i.e., `ceph::condition_variable_debug`, which works with `std::unique_lock<ceph::mutex_debug>`. so if we put something like
```c++
ceph::mutex m;
ceph::condition_variable cv;
// ...
std::unique_lock lock{m};
cv.wait(lock);
```

it will work as expected.

if we do intend to use `std::condition_variable<>` along with `ceph::mutex`, i don't think it works. because if `CEPH_DEBUG_MUTEX` is defined. what we will have is `ceph::mutex_debug`, which is not a `std::mutex`. while what `std::condition_variable::wait()` expects is `std::unique_lock<std::mutex>`.

Reply:
i see, we need to keep using `ceph::condition_variable` as an alias of `std::condition_variable` in the "release" mode. but template deduction will always consider that the `m` in `std::unique_lock m{some_ceph_mutex}` a `std::unique_lock<ceph::mutex>` which unacceptable from `std::condition_variable` 's POV.

Reply:
i think we should instead add a deduction guide for `std::unique_lock`, like

```c++
namespace std {
  unique_lock(ceph::mutex&) -> unique_lock<std::mutex>;
}
```

to make sure the template deduction do the right thing, and without forcing `ceph::unique_lock` to be an instantiated template, which is not intuitive and surprising.

Reply:
Yeah, currently the glue that makes this work is
```
    // enable silent conversion to std::mutex, e.g. for lock_guard<>
    operator std::mutex&() {
      return m;
    }
```
Let me try it the deduction guide... that would be way better!

Reply:
note that http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4659.pdf#paragraph.20.5.4.2.1 says:
> The behavior of a C++ program is undefined if it declares ... a deduction guide for any standard library class template.

Reply:
sorry i'm not following why @tchaikov's example above doesn't work. if ceph::mutex is an alias of std::mutex, then `std::unique_lock<ceph::mutex>` is the same thing as `std::unique_lock<std::mutex>`, and should work fine with std::condition_variable

Reply:
It works!

Reply:
> The behavior of a C++ program is undefined if it declares ... a deduction guide for any standard library class template.

that's also my assumption... actually, when i was typing `namespace std {` in that comment,  i hoped that you or Adam wouldn't notice it.  i do want to respect the C++17 standard. but when facing the temptation of a simpler (and probably viable) solution, the inner pragmatist me beats the standard-compliant me. probably we can make an exception? we can always switch back to plan B, if something bad happens when, for example, we upgrade the libstd++ or compiler?

Reply:
That makes sense to me.  It seems like this is unlikely to break in practice.. and if it does we can fall back to sed 's/std::unique_lock/ceph::unique_lock/'.

Reply:
> actually, when i was typing namespace std { in that comment, i hoped that you or Adam wouldn't notice it

mwahaha

i agree that it's extremely unlikely to break anything, but it might be worth a comment around the deduction guides :+1: